apps/ejabberd/src/mod_register.erl

```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
-								   Server,
-								   Password) of
-					ok ->
-					    IQ#iq{type = result,
-						  sub_el = [SubEl]};
-					%% TODO FIXME: This piece of
-					%% code does not work since
-					%% the code have been changed
-					%% to allow several auth
-					%% modules.  lists:foreach can
-					%% only return ok:
-					not_allowed ->
-					    IQ#iq{type = error,
-						  sub_el =
-						  [SubEl, ?ERR_NOT_ALLOWED]};
-					not_exists ->
-					    IQ#iq{type = error,
-						  sub_el =
-						  [SubEl, ?ERR_ITEM_NOT_FOUND]};
-					_ ->
-					    IQ#iq{type = error,
-						  sub_el =
-						  [SubEl,
-						   ?ERR_INTERNAL_SERVER_ERROR]}
-				    end;
-				true ->
-				    IQ#iq{type = error,
-					  sub_el = [SubEl, ?ERR_BAD_REQUEST]}
-			    end
-		    end;
-		(UTag == false) and (RTag /= false) and AllowRemove ->
-		    case From of
-			#jid{user = User,
-			     lserver = Server,
-			     resource = Resource} ->
-			    ResIQ = #iq{type = result, xmlns = ?NS_REGISTER,
-					id = ID,
-					sub_el = [SubEl]},
-			    ejabberd_router:route(
-			      jid:make(User, Server, Resource),
-			      jid:make(User, Server, Resource),
-			      jlib:iq_to_xml(ResIQ)),
-			    ejabberd_auth:remove_user(User, Server),
-			    ignore;
-			_ ->
-			    IQ#iq{type = error,
-				  sub_el = [SubEl, ?ERR_NOT_ALLOWED]}
-		    end;
-		(UTag /= false) and (PTag /= false) ->
-		    User = xml:get_tag_cdata(UTag),
-		    Password = xml:get_tag_cdata(PTag),
-		    try_register_or_set_password(
-		      User, Server, Password, From,
-		      IQ, SubEl, Source, Lang);
-		true ->
-		    IQ#iq{type = error,
-			  sub_el = [SubEl, ?ERR_BAD_REQUEST]}
-	    end;
-	get ->
-	    {_IsRegistered, UsernameSubels, QuerySubels} =
-		case From of
-		    #jid{user = User, lserver = Server} ->
-			case ejabberd_auth:is_user_exists(User, Server) of
-			    true ->
-				{true, [#xmlcdata{content = User}],
-				 [#xmlel{name = <<"registered">>}]};
-			    false ->
-				{false, [#xmlcdata{content = User}], []}
-			end;
-		    _ ->
-			{false, [], []}
-		end,
-        TranslatedMsg = translate:translate(
-                          Lang, <<"Choose a username and password to register with this server">>),
-		    IQ#iq{type = result,
-		  sub_el = [#xmlel{name = <<"query">>,
-				   attrs = [{<<"xmlns">>, <<"jabber:iq:register">>}],
-				   children = [#xmlel{name = <<"instructions">>,
-				              children = [#xmlcdata{content = TranslatedMsg}]},
-				       #xmlel{name = <<"username">>,
-					      children = UsernameSubels},
-				       #xmlel{name = <<"password">>}
-				       | QuerySubels]}]}
+%% Clients must register before being able to authenticate.
+process_unauthenticated_iq(From, To, #iq{type = set} = IQ, IPAddr) ->
+    process_iq_set(From, To, IQ, IPAddr);
+process_unauthenticated_iq(From, To, #iq{type = get} = IQ, IPAddr) ->
+    process_iq_get(From, To, IQ, IPAddr).
+
+process_iq(From, To, #iq{type = set} = IQ) ->
+    process_iq_set(From, To, IQ, jlib:jid_tolower(From));
```

Comment:
Please take a look at `jid` module and change all the deprecated functions from `jlib`.